server/src/main/java/org/opensearch/common/blobstore/EncryptedBlobStore.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore;
+
+import org.opensearch.cluster.metadata.CryptoMetadata;
+import org.opensearch.crypto.CryptoManager;
+import org.opensearch.crypto.CryptoManagerRegistry;
+import org.opensearch.crypto.CryptoRegistryException;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * The EncryptedBlobStore is a decorator class that wraps an existing BlobStore and provides encryption and decryption
+ * capabilities for the stored data. It uses a CryptoManager to handle encryption and decryption operations based on
+ * the provided CryptoMetadata. The EncryptedBlobStore ensures that all data written to and read from the underlying
+ * BlobStore is encrypted and decrypted transparently.
+ */
+public class EncryptedBlobStore implements BlobStore {
+
+    private final BlobStore blobStore;
+    private final CryptoManager cryptoManager;
+
+    /**
+     * Constructs an EncryptedBlobStore that wraps the provided BlobStore with encryption capabilities based on the
+     * given CryptoMetadata.
+     *
+     * @param blobStore     The underlying BlobStore to be wrapped and used for storing encrypted data.
+     * @param cryptoMetadata The CryptoMetadata containing information about the key provider and settings for encryption.
+     * @throws CryptoRegistryException If the CryptoManager is not found during encrypted BlobStore creation.
+     */
+    public EncryptedBlobStore(BlobStore blobStore, CryptoMetadata cryptoMetadata) {
+        this.cryptoManager = CryptoManagerRegistry.fetchCryptoManager(cryptoMetadata);
+        if (cryptoManager == null) {
+            throw new CryptoRegistryException(
+                cryptoMetadata.keyProviderName(),
+                cryptoMetadata.keyProviderType(),
+                "Crypto manager not found during encrypted blob store creation."
+            );
+        }
+        this.cryptoManager.incRef();
+        this.blobStore = blobStore;
+    }
+
```

Comment:
Do we need to persist crypto metadata like frame length, encryption algorithm, version for BWC as a part of the repository metadata blob to ensure we don't have to duplicate the same effort of maintaining this metadata at a per file/block basis. This should help with lesser marginal computation overhead and perf benefits